src/lib.rs

```
@@ -0,0 +1,277 @@
+    ///
+    /// let mut rng = rand::thread_rng();
+    /// let num_messages = 5;
+    /// let params = BbsParams::new(num_messages, &mut rng);
+    /// let x = Scalar::random(&mut rng);
+    /// let mut messages = Vec::with_capacity(num_messages);
+    /// for _ in 0..num_messages {
+    ///     messages.push(Scalar::random(&mut rng));
+    /// }
+    ///
+    /// let mac = BbsMac::sign(&x, &messages, &params, &mut rng).unwrap();
+    /// BbsMac::verify(&x, &messages, &mac, &params).unwrap();
+    /// ```
+    pub fn sign(
+        x: &Scalar,
+        messages: &[Scalar],
+        bbs_params: &BbsParams<G1Projective>,
+        rng: &mut impl RngCore,
+    ) -> Result<BbsMac, BbsMacError> {
+        if messages.len() != bbs_params.h_generators.len() {
+            return Err(BbsMacError::InputLengthMismatch);
+        }
+
+        // generate random e for the MAC
+        let mut e: Scalar = Scalar::random(&mut *rng);
+
+        // Compute the scalar inverse: (x + e)^-1.
+        let mut d: Scalar = *x + e; // d = x + e
+        // We loop, re-sampling 'e' until (x + e) is non-zero and we get an inverse.
+        let d_inv: Scalar = loop {
+            // Attempt to compute the inverse.
+            // ff::Field::invert returns subtle::CtOption<Self>.
+            let inv = d.invert();
+            if bool::from(inv.is_some()) {
+                // Success! The inverse exists.
+                break inv.unwrap();
+            } else {
+                // (x + e) was zero. Sample a new 'e' and the loop will retry.
+                e = Scalar::random(&mut *rng);
+                d = *x + e;
+            }
+        };
+
+        // Compute the multi-scalar multiplication with the generators and messages:
+        // `B = G + m1*H1 + ... + ml*Hl`
+        // TODO do we have some form of multi-exp here?
+        let b: G1Projective = bbs_params.g_generator
+            + messages
+                .iter()
+                .zip(bbs_params.h_generators.iter())
```

Something to be aware of is that if you zip iterators of different lengths, the longer iterator will get silently truncated.

---

good to know, thanks! we check it at the beginning of the function so it should be fine.